File: packages/cli-go/internal/scanner/scanner.go
package scanner

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// Default ignore patterns
var defaultIgnore = []string{
	"node_modules", ".git", ".next", "dist", "build",
	"__pycache__", ".venv", "vendor", ".idea", ".vscode",
	"coverage", ".nyc_output", ".turbo", ".cache",
}

// Source file extensions we care about
var sourceExtensions = map[string]string{
	".ts":     "typescript",
	".tsx":    "typescript-react",
	".js":     "javascript",
	".jsx":    "javascript-react",
	".py":     "python",
	".go":     "go",
	".rs":     "rust",
	".java":   "java",
	".rb":     "ruby",
	".php":    "php",
	".swift":  "swift",
	".kt":     "kotlin",
	".cs":     "csharp",
	".cpp":    "cpp",
	".c":      "c",
	".h":      "c-header",
	".vue":    "vue",
	".svelte": "svelte",
}

// ScanResult represents a scanned directory's context.
type ScanResult struct {
	FeatureName string
	Content     string
	EntryType   string
	Tags        []string
}

// ScanOptions configures the scanner.
type ScanOptions struct {
	Path   string
	Depth  int
	Ignore []string
}

// ScanDirectory scans a directory and produces context entries.
func ScanDirectory(opts ScanOptions) (files []string, results []ScanResult, err error) {
	root, err := filepath.Abs(opts.Path)
	if err != nil {
		return nil, nil, fmt.Errorf("resolve path: %w", err)
	}

	ignoreSet := make(map[string]bool)
	for _, p := range defaultIgnore {
		ignoreSet[p] = true
	}
	for _, p := range opts.Ignore {
		ignoreSet[p] = true
	}

	// Group files by their parent directory
	dirFiles := make(map[string][]string)

	err = filepath.Walk(root, func(path string, info os.FileInfo, walkErr error) error {
		if walkErr != nil {
			return nil // Skip errors
		}

		rel, _ := filepath.Rel(root, path)

		// Check depth
		depth := strings.Count(rel, string(filepath.Separator))
		if depth > opts.Depth {
			if info.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}

		// Skip ignored directories
		if info.IsDir() {
			if ignoreSet[info.Name()] {
				return filepath.SkipDir
			}
			return nil
		}

		// Only process source files
		ext := strings.ToLower(filepath.Ext(info.Name()))
		if _, ok := sourceExtensions[ext]; !ok {
			return nil
		}

		files = append(files, path)
		dir := filepath.Dir(rel)
		if dir == "." {
			dir = filepath.Base(root)
		}
		dirFiles[dir] = append(dirFiles[dir], rel)

		return nil
	})
	if err != nil {
		return nil, nil, fmt.Errorf("scan directory: %w", err)
	}

	// Build context entries per directory
	for dir, dirFileList := range dirFiles {
		featureName := strings.ReplaceAll(dir, string(filepath.Separator), "/")

		var contentBuilder strings.Builder
		contentBuilder.WriteString(fmt.Sprintf("Auto-scanned directory: %s\n\n", dir))
		contentBuilder.WriteString(fmt.Sprintf("Files (%d):\n", len(dirFileList)))

		langCounts := make(map[string]int)
		for _, f := range dirFileList {
			ext := strings.ToLower(filepath.Ext(f))
			if lang, ok := sourceExtensions[ext]; ok {
				langCounts[lang]++
			}
			contentBuilder.WriteString(fmt.Sprintf("  - %s\n", f))
		}

		contentBuilder.WriteString("\nLanguages:\n")
		for lang, count := range langCounts {
			contentBuilder.WriteString(fmt.Sprintf("  - %s: %d files\n", lang, count))
		}

		tags := []string{"auto-scan"}
		for lang := range langCounts {
			tags = append(tags, lang)
		}

		results = append(results, ScanResult{
			FeatureName: featureName,
			Content:     contentBuilder.String(),
			EntryType:   "scan",
			Tags:        tags,
		})
	}

	return files, results, nil
}

File: packages/cli/src/lib/scanner.ts
import { readFileSync, statSync, existsSync } from "node:fs";
import { resolve, relative, extname, basename, dirname } from "node:path";
import { glob } from "glob";

/**
 * Default ignore patterns for scanning.
 */
const DEFAULT_IGNORE = [
  "**/node_modules/**",
  "**/.git/**",
  "**/dist/**",
  "**/build/**",
  "**/.next/**",
  "**/.nuxt/**",
  "**/.svelte-kit/**",
  "**/coverage/**",
  "**/__pycache__/**",
  "**/.venv/**",
  "**/venv/**",
  "**/*.min.js",
  "**/*.min.css",
  "**/*.map",
  "**/*.lock",
  "**/pnpm-lock.yaml",
  "**/package-lock.json",
  "**/yarn.lock",
  "**/.env*",
  "**/*.png",
  "**/*.jpg",
  "**/*.jpeg",
  "**/*.gif",
  "**/*.svg",
  "**/*.ico",
  "**/*.woff*",
  "**/*.ttf",
  "**/*.eot",
];

/**
 * File extensions we consider scannable source code.
 */
const SOURCE_EXTENSIONS = new Set([
  ".ts",
  ".tsx",
  ".js",
  ".jsx",
  ".mjs",
  ".mts",
  ".py",
  ".go",
  ".rs",
  ".java",
  ".kt",
  ".rb",
  ".php",
  ".swift",
  ".c",
  ".cpp",
  ".h",
  ".cs",
  ".vue",
  ".svelte",
  ".astro",
  ".md",
  ".mdx",
  ".yaml",
  ".yml",
  ".toml",
  ".ini",
  ".cfg",
  ".sql",
  ".graphql",
  ".gql",
  ".prisma",
  ".proto",
  ".tf",
  ".dockerfile",
  ".sh",
  ".bash",
  ".zsh",
]);

export interface ScannedFile {
  relativePath: string;
  language: string;
  size: number;
  lines: number;
  content: string;
}

export interface ScanResult {
  featureName: string;
  content: string;
  entryType: string;
  tags: string[];
}

interface ScanOptions {
  path: string;
  depth?: number;
  ignore?: string[];
  maxFileSize?: number;
}

/**
 * Scan a directory and return structured context entries.
 */
export async function scanDirectory(
  opts: ScanOptions
): Promise<{ files: ScannedFile[]; results: ScanResult[] }> {
  const {
    path: scanPath,
    depth = 5,
    ignore = [],
    maxFileSize = 100_000, // 100KB per file
  } = opts;

  const rootDir = resolve(scanPath);

  if (!existsSync(rootDir) || !statSync(rootDir).isDirectory()) {
    throw new Error(`Directory not found: ${rootDir}`);
  }

  const allIgnore = [...DEFAULT_IGNORE, ...ignore];

  const files = await glob("**/*", {
    cwd: rootDir,
    nodir: true,
    ignore: allIgnore,
    maxDepth: depth,
    absolute: false,
  });

  const scannedFiles: ScannedFile[] = [];

  for (const file of files) {
    const ext = extname(file).toLowerCase();
    if (!SOURCE_EXTENSIONS.has(ext) && ext !== "") continue;

    const fullPath = resolve(rootDir, file);
    const stat = statSync(fullPath);

    if (stat.size > maxFileSize || stat.size === 0) continue;

    const raw = readFileSync(fullPath, "utf-8");
    scannedFiles.push({
      relativePath: file,
      language: extToLanguage(ext),
      size: stat.size,
      lines: raw.split("\n").length,
      content: raw,
    });
  }

  // Group files by directory (each directory ≈ a feature)
  const dirGroups = new Map<string, ScannedFile[]>();
  for (const f of scannedFiles) {
    const dir = dirname(f.relativePath);
    const group = dirGroups.get(dir) ?? [];
    group.push(f);
    dirGroups.set(dir, group);
  }

  const results: ScanResult[] = [];

  for (const [dir, groupFiles] of dirGroups) {
    const featureName = dir === "." ? basename(rootDir) : dir.replace(/\//g, "-");

    // Build a summary of the directory
    const fileSummaries = groupFiles.map((f) => {
      const preview =
        f.content.length > 500
          ? f.content.slice(0, 500) + "\n... (truncated)"
          : f.content;
      return `### ${f.relativePath}\n\`\`\`${f.language}\n${preview}\n\`\`\``;
    });

    const entryContent = [
      `# Directory: ${dir === "." ? basename(rootDir) : dir}`,
      ``,
      `**Files:** ${groupFiles.length} | **Languages:** ${[...new Set(groupFiles.map((f) => f.language))].join(", ")}`,
      `**Total lines:** ${groupFiles.reduce((sum, f) => sum + f.lines, 0)}`,
      ``,
      ...fileSummaries,
    ].join("\n");

    // Avoid exceeding the API's 50k char limit
    const truncated =
      entryContent.length > 45_000
        ? entryContent.slice(0, 45_000) + "\n\n... (truncated due to size)"
        : entryContent;

    results.push({
      featureName,
      content: truncated,
      entryType: "scan",
      tags: [
        "auto-scan",
        ...new Set(groupFiles.map((f) => f.language)),
      ],
    });
  }

  return { files: scannedFiles, results };
}

function extToLanguage(ext: string): string {
  const map: Record<string, string> = {
    ".ts": "typescript",
    ".tsx": "tsx",
    ".js": "javascript",
    ".jsx": "jsx",
    ".mjs": "javascript",
    ".mts": "typescript",
    ".py": "python",
    ".go": "go",
    ".rs": "rust",
    ".java": "java",
    ".kt": "kotlin",
    ".rb": "ruby",
    ".php": "php",
    ".swift": "swift",
    ".c": "c",
    ".cpp": "cpp",
    ".h": "c",
    ".cs": "csharp",
    ".vue": "vue",
    ".svelte": "svelte",
    ".astro": "astro",
    ".md": "markdown",
    ".mdx": "mdx",
    ".yaml": "yaml",
    ".yml": "yaml",
    ".toml": "toml",
    ".sql": "sql",
    ".graphql": "graphql",
    ".gql": "graphql",
    ".prisma": "prisma",
    ".proto": "protobuf",
    ".tf": "terraform",
    ".sh": "bash",
    ".bash": "bash",
    ".zsh": "zsh",
    ".dockerfile": "dockerfile",
    ".ini": "ini",
    ".cfg": "ini",
  };
  return map[ext] ?? "text";
}
